feat: add json data type

CookiePieWw opened the pull request:
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#3686 #4515 #4230

## What's changed and what's your intention?

As title. Currently the datatype can store json strings as jsonb format. ~~Also a simple udf is added to query json through a path.~~ Now the udf is deleted. We can focus on the type first and add some udfs in the other PRs.

e.g.
![image](https://github.com/user-attachments/assets/0c3681c4-6b94-47a6-a619-52d8d9ab3e82)

Since the underlying storage of json is the same as binary, we cannot distinguish them inside datafusion, thus a comment is attached to the metadata of `Field` of json column for frontend to identify and convert jsonb to strings.

TODO:
Unit tests and sqlness tests.
UDFs.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced support for JSON data types throughout the application, including new functions for JSON handling.
	- Added a `JsonFunction` for enhanced JSON operations.
	- Expanded the functionality of the `FUNCTION_REGISTRY` to include JSON operations.
- **Improvements**
	- Enhanced data handling for JSON in SQL parsing and data encoding across various components.
	- Updated data equality checks to support JSON types.
- **Chores**
	- Added new dependencies related to JSON handling in various configuration files.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

WenyXu commented:
The CI is failed, try `make fmt-toml`

WenyXu commented on Cargo.toml:
```diff
@@ -120,10 +120,11 @@ etcd-client = { version = "0.13" }
 humantime = "2.1"
 humantime-serde = "1.1"
 itertools = "0.10"
+jsonb = "0.4.1"
```
```suggestion
jsonb = "0.4"
```

WenyXu commented on tests/cases/standalone/common/insert/merge_mode.result:
```diff
@@ -103,7 +103,7 @@ create table if not exists invalid_merge_mode(
 with('merge_mode'='first_row');
 
-Error: 1004(InvalidArguments), Invalid options: Matching variant not found at line 1 column 25
+Error: 1004(InvalidArguments), Invalid options: Matching variant not found at line 1 column 76
```
Let's use the magic.

tests/cases/standalone/common/insert/merge_mode.sql#L47
```sql
-- SQLNESS REPLACE (line\s\d+\scolumn\s\d+) PLACE
create table if not exists invalid_merge_mode(
    host string,
    ts timestamp,
    cpu double,
    memory double,
    TIME INDEX (ts),
    PRIMARY KEY(host)
)
engine=mito
with('merge_mode'='first_row');
```

WenyXu commented:
@coderabbitai review

killme2008 commented:
A great work! 
I don't like the name `get_by_path`, it's too long. What about `jq`? just like the command line tool https://jqlang.github.io/jq/

And i found that databend is using this name too.

CookiePieWw commented:
> What about jq? just like the command line tool https://jqlang.github.io/jq/

Sure, it's a good name if there are already some conventions :)

WenyXu commented on src/datatypes/src/value.rs:
```diff
@@ -77,6 +77,8 @@ pub enum Value {
 
     List(ListValue),
+
+    Json(Bytes),
```
I think we can remove the type here. We can considered the JSON as kind of logical type.

WenyXu commented on src/common/function/src/scalars/json/jq.rs:
We can put udf support into a separate PR. 

i.e.,
```sql
select get_by_path_int(parse_jsonb(''[{"a":1}, {"a":2}]''),'[:1].a');
```

WenyXu commented on src/datatypes/src/schema.rs:
```diff
@@ -256,7 +256,15 @@ fn collect_fields(column_schemas: &[ColumnSchema]) -> Result<FieldsAndIndices> {
+        let mut field = Field::try_from(column_schema)?;
+
+        // Json column performs the same as binary column in Arrow, so we need to mark it
+        if column_schema.data_type.is_json() {
+            field.set_metadata(HashMap::from([(
+                String::from("is_json"),
+                String::from("true"),
+            )]));
+        }
```
Add unit tests, ensures other people won't break your feature.

WenyXu reviewed (commented):
Can we add sqlness test to ensure basic write/query behavior of JSON data type? BTW, mysql, pg write/query integration tests are also important, but they can be added in future PRs.

killme2008 commented on Cargo.toml:
```diff
@@ -124,6 +124,7 @@ greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", r
 humantime = "2.1"
 humantime-serde = "1.1"
 itertools = "0.10"
+jsonb = { git = "https://github.com/CookiePieWw/jsonb.git", branch = "trigger-preserve-order-feature" }
```
Is it a temporary repository?

CookiePieWw replied:
Yes it's temporary. `jsonb` conflicts with greptimedb for `serde_json`. I've open an issue for it (https://github.com/datafuselabs/jsonb/issues/55), if the change works for them then we can use their repo, or I may find  other repo for jsonb format.

killme2008 commented on src/common/function/src/scalars/json/jq.rs:
```diff
@@ -0,0 +1,174 @@
+
+use crate::function::{Function, FunctionContext};
+
+#[derive(Clone, Debug, Default)]
```
```suggestion
/// A function to retrieve the value in JSON by path.
#[derive(Clone, Debug, Default)]
```

killme2008 replied:
`jq` may be not a good name, sorry for misunderstanding. jq is more complex than this function. It can transform, filter the JSON.

There are some JSON functions or operators in databend

https://github.com/datafuselabs/databend/issues/11270

killme2008 commented on src/common/function/src/scalars/json/jq.rs:
```diff
@@ -0,0 +1,174 @@
+        match datatype {
+            ConcreteDataType::Binary(_) | ConcreteDataType::Json(_) => {
+                for i in 0..size {
+                    let json = jsons.get_ref(i);
```
Is it safe to `unwrap` here?

killme2008 commented on src/datatypes/src/schema.rs:
```diff
@@ -256,7 +256,15 @@ fn collect_fields(column_schemas: &[ColumnSchema]) -> Result<FieldsAndIndices> {
+        // Json column performs the same as binary column in Arrow, so we need to mark it
+        if column_schema.data_type.is_json() {
+            field.set_metadata(HashMap::from([(
+                String::from("is_json"),
```
Make `is_json` key to be a constant.

killme2008 commented on src/datatypes/src/types/json_type.rs:
```diff
@@ -0,0 +1,63 @@
+use crate::vectors::{BinaryVectorBuilder, MutableVector};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct JsonType;
```
Please always add comments for pub structs.

killme2008 commented on src/mito2/src/row_converter.rs:
```diff
@@ -192,6 +193,10 @@ impl SortField {
                         data_type: ConcreteDataType::Null(n.clone()),
                     }
                     .fail(),
+                    ConcreteDataType::Json(j) => NotSupportedFieldSnafu {
```
I think we should process it just like binary type here.

killme2008 commented on src/query/src/dist_plan/merge_scan.rs:
```diff
@@ -156,20 +156,19 @@ impl MergeScanExec {
         target_partition: usize,
     ) -> Result<Self> {
-        let arrow_schema_without_metadata = Self::arrow_schema_without_metadata(arrow_schema);
+        let arrow_schema = Arc::new(arrow_schema.clone());
```
The logic is changed here. Please confirm it. @waynexia

killme2008 commented on src/servers/src/mysql/writer.rs:
```diff
@@ -208,7 +210,14 @@ impl<'a, W: AsyncWrite + Unpin> MysqlResultWriter<'a, W> {
                     Value::Float64(v) => row_writer.write_col(v.0)?,
                     Value::String(v) => row_writer.write_col(v.as_utf8())?,
-                    Value::Binary(v) => row_writer.write_col(v.deref())?,
+                    Value::Binary(v) | Value::Json(v) => match column.coltype {
```
I'd like to keep the binary branch not changed, but add a new match branch:

```rust
Value::Json(v) if column.coltype == ColumnType::MYSQL_TYPE_JSON => {
    ......
}
```

killme2008 commented on src/servers/src/postgres/types.rs:
```diff
@@ -77,13 +78,18 @@ pub(super) fn encode_value(
-            match *bytea_output {
-                PGByteaOutputValue::ESCAPE => builder.encode_field(&EscapeOutputBytea(v.deref())),
-                PGByteaOutputValue::HEX => builder.encode_field(&HexOutputBytea(v.deref())),
+        Value::Binary(v) | Value::Json(v) => match datatype {
```
The same suggestion as above.

killme2008 commented on src/sql/src/statements.rs:
```diff
@@ -124,6 +124,16 @@ fn parse_string_to_value(
             }
         }
         ConcreteDataType::Binary(_) => Ok(Value::Binary(s.as_bytes().into())),
+        ConcreteDataType::Json(_) => {
```
Need a unit test.

killme2008 reviewed (commented):
Some suggestions attached. And we need the sqlness test cases to cover this type, it can be in another PR.

CookiePieWw commented:
@WenyXu @killme2008 Hi, I've resolved some bugs and made some changes, PTAL :)

Changes:
1. Added a type extension in proto to identify the json type from binary type in proto messages. See https://github.com/GreptimeTeam/greptime-proto/pull/188 for changes in proto.
2. Removed the unused `Value(Json)` and made use of current `Value(Binary)` as the underlying representation of values of json type.
3. Deleted the udf. Leave it with future PRs about udfs.
4. Prohibited to directly insert into json with hex strings e.g. X'abcd'. Now it should only accept plain strings.
5. Fixed and added some unit tests.

WenyXu commented on src/datatypes/src/value.rs:
```diff
@@ -1037,7 +1037,18 @@ impl<'a> ValueRef<'a> {
     /// Cast itself to binary slice.
     pub fn as_binary(&self) -> Result<Option<&[u8]>> {
-        impl_as_for_value_ref!(self, Binary)
+        match self {
+            ValueRef::Null => Ok(None),
+            ValueRef::Binary(v) => Ok(Some(*v)),
+            other => error::CastTypeSnafu {
+                msg: format!(
+                    "Failed to cast value ref {:?} to {}",
+                    other,
+                    stringify!(Binary)
+                ),
+            }
+            .fail(),
+        }
```
```suggestion
        impl_as_for_value_ref!(self, Binary)
```

WenyXu commented on src/datatypes/src/schema/column_schema.rs:
```diff
@@ -19,6 +19,7 @@ use arrow::datatypes::Field;
 use serde::{Deserialize, Serialize};
 use snafu::{ensure, ResultExt};
 
+use super::TYPE_KEY;
```
```suggestion
use crate::schema::TYPE_KEY;
```

WenyXu commented on src/servers/src/grpc/flight/stream.rs:
```diff
@@ -129,6 +157,54 @@ impl Stream for FlightRecordBatchStream {
+
+/// Column of json type in record batches needs to be serialized before sending
+/// to the client. This function serializes the binary data of json type to string.
+fn serialize_json_column(schema: Arc<Schema>, recordbatch: &RecordBatch) -> RecordBatch {
+    let mut column_schemas = Vec::with_capacity(schema.column_schemas().len());
+    let mut vectors: Vec<Arc<dyn Vector>> = Vec::with_capacity(schema.column_schemas().len());
+    for (i, column) in schema.column_schemas().iter().enumerate() {
+        let vector = recordbatch.column(i);
+        if column.data_type.is_json() {
+            // Safety: json column with binary vector
+            let vector = vector.as_any().downcast_ref::<BinaryVector>().unwrap();
+            let vector = vector.from_json_vector_to_string_vector();
+            vectors.push(Arc::new(vector) as VectorRef);
+            column_schemas.push(ColumnSchema::new(
+                column.name.clone(),
+                ConcreteDataType::string_datatype(),
+                column.is_nullable(),
+            ));
+        } else {
+            vectors.push(vector.clone());
+            column_schemas.push(column.clone());
+        }
+    }
+
+    // Safety: only changed a column type from binary to string
+    RecordBatch::new(Arc::new(Schema::new(column_schemas)), vectors).unwrap()
+}
```
How about moving this part into a UDF, we only use the grpc protocol for internal usage(e.g., frontend <-> datanode, sqlness tests). we won't use grpc protocol for queries.

WenyXu commented on src/query/src/dist_plan/merge_scan.rs:
```diff
@@ -156,20 +156,19 @@ impl MergeScanExec {
-            EquivalenceProperties::new(arrow_schema_without_metadata.clone()),
+            EquivalenceProperties::new(arrow_schema.clone()),
             Partitioning::UnknownPartitioning(target_partition),
             ExecutionMode::Bounded,
         );
-        let schema_without_metadata =
-            Self::arrow_schema_to_schema(arrow_schema_without_metadata.clone())?;
+        let schema = Self::arrow_schema_to_schema(arrow_schema.clone())?;
         Ok(Self {
             table,
             regions,
             plan,
-            schema: schema_without_metadata,
-            arrow_schema: arrow_schema_without_metadata,
+            schema,
+            arrow_schema,
             region_query_handler,
             metric: ExecutionPlanMetricsSet::new(),
             sub_stage_metrics: Arc::default(),
```
Let's introduce a `remove_column_metadata` method for `Schema`

```rust
    pub fn remove_column_metadata(mut self) -> Schema {
        self.column_schemas = self
            .column_schemas
            .into_iter()
            .map(|column_schema| column_schema.with_metadata(Default::default()))
            .collect();

        self.arrow_schema = Arc::new(ArrowSchema::new(
            self.arrow_schema
                .fields()
                .iter()
                .map(|field| {
                    let field = field.as_ref().clone();
                    let field_without_metadata = field.with_metadata(Default::default());
                    Arc::new(field_without_metadata)
                })
                .collect::<Vec<_>>(),
        ));

        self
    }
```

then you can use it like: 

```rust
let schema = Self::arrow_schema_to_schema(arrow_schema.clone())?.remove_column_schema();
let arrow_schema = schema.arrow_schema().clone();

...
```

discord9 commented on src/datatypes/src/schema/column_schema.rs:
```diff
@@ -34,6 +35,8 @@ pub const COMMENT_KEY: &str = "greptime:storage:comment";
 /// Key used to store fulltext options in arrow field's metadata.
 pub const FULLTEXT_KEY: &str = "greptime:fulltext";
+/// Value used to identify column data type from arrow field's metadata. It should be the same as `ConcreteDataType::json_datatype().name()`.
+pub const JSON_TYPE_VALUE: &str = "Json";
```
nitpick: use the same const in both here and json type def(since in same crate anyway)?

CookiePieWw replied:
Reasonable to me :)

discord9 commented on tests/cases/standalone/common/types/json/json.sql:
```diff
@@ -0,0 +1,32 @@
+
+INSERT INTO jsons VALUES('\xaa\xff\xaa'::BYTEA, 1), ('\xaa\xff\xaa\xaa\xff\xaa'::BYTEA, 2), ('\xaa\xff\xaa\xaa\xff\xaa\xaa\xff\xaa'::BYTEA, 3);
+
+SELECT * FROM blobs;
```
is this `blobs` supposed to be `jsons`?

CookiePieWw replied:
Yes, I misspelled here qaq. Changed already (actually removed this test owo)

discord9 commented:
Rest LGTM

WenyXu commented on src/common/function/src/scalars/json/json_to_string.rs:
```diff
@@ -0,0 +1,173 @@
+
+use crate::function::{Function, FunctionContext};
+
+#[derive(Clone, Debug, Default)]
+pub struct JsonToStringFunction;
```
```suggestion
/// Converts the `JSONB` into `String`. It's useful for displaying JSONB content.
#[derive(Clone, Debug, Default)]
pub struct JsonToStringFunction;
```

WenyXu commented on src/common/function/src/scalars/json/to_json.rs:
```diff
@@ -0,0 +1,164 @@
+
+use crate::function::{Function, FunctionContext};
+
+#[derive(Clone, Debug, Default)]
+pub struct ToJsonFunction;
```
```suggestion
/// Parses the `String` into `JSONB`.
#derive(Clone, Debug, Default)]
pub struct ToJsonFunction;
```

WenyXu commented on src/datatypes/src/vectors/binary.rs:
```diff
@@ -36,6 +38,16 @@ impl BinaryVector {
         &self.array
     }
+
+    // Create a StringVector from a BinaryVector of json type
+    pub fn from_json_vector_to_string_vector(&self) -> StringVector {
+        let mut builder = StringVectorBuilder::with_capacity(self.len());
+        for i in 0..self.len() {
+            let value = self.get_data(i).map(jsonb::to_string);
+            builder.push(value.as_deref());
+        }
+        builder.finish()
+    }
```
Let's remove the dead code

WenyXu commented on src/servers/src/grpc/flight/stream.rs:
```diff
@@ -128,7 +128,6 @@ impl Stream for FlightRecordBatchStream {
         }
     }
 }
-
```
No need to modify

WenyXu reviewed (commented):
Just some nitpicking

WenyXu commented:
Hi @CookiePieWw, there are some conflicts

WenyXu commented on Cargo.toml:
```diff
@@ -120,10 +120,11 @@ etcd-client = { version = "0.13" }
 futures = "0.3"
 futures-util = "0.3"
-greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "157cfdb52709e489cf1f3ce8e3042ed4ee8a524a" }
+greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "9d5c8a9e0f7a283bcd26742faeb636e885ca1ec5" }
```
```suggestion
greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "973f49cde88a582fb65755cc572ebcf6fb93ccf7" }
```

WenyXu commented on Cargo.lock:
```diff
@@ -4303,7 +4311,7 @@ checksum = "d2fabcfbdc87f4758337ca535fb41a6d701b65693ce38287d856d1674551ec9b"
 name = "greptime-proto"
 version = "0.1.0"
-source = "git+https://github.com/GreptimeTeam/greptime-proto.git?rev=157cfdb52709e489cf1f3ce8e3042ed4ee8a524a#157cfdb52709e489cf1f3ce8e3042ed4ee8a524a"
+source = "git+https://github.com/GreptimeTeam/greptime-proto.git?rev=9d5c8a9e0f7a283bcd26742faeb636e885ca1ec5#9d5c8a9e0f7a283bcd26742faeb636e885ca1ec5"
```
```suggestion
source = "git+https://github.com/GreptimeTeam/greptime-proto.git?rev=973f49cde88a582fb65755cc572ebcf6fb93ccf7#973f49cde88a582fb65755cc572ebcf6fb93ccf7"
```

paomian commented:
Nice job!
